cmd/src/users_clean.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/jedib0t/go-pretty/v6/table"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+)
+
+func init() {
+	usage := `
+This command removes users from a Sourcegraph instance who have been inactive for 60 or more days.
+	
+Examples:
+
+	$ src users clean -days 182
+	
+	$ src users clean -noAdmin -removeNeverActive 
+`
+
+	flagSet := flag.NewFlagSet("clean", flag.ExitOnError)
+	usageFunc := func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "Usage of 'src users %s':\n", flagSet.Name())
+		flagSet.PrintDefaults()
+		fmt.Println(usage)
+	}
+	var (
+		daysToDelete       = flagSet.Int("days", 60, "Days threshold on which to remove users, must be 60 days or greater and defaults to this value ")
+		noAdmin            = flagSet.Bool("noAdmin", false, "Omit admin accounts from cleanup")
+		removeNoLastActive = flagSet.Bool("removeNeverActive", false, "removes users with null lastActive value")
+		skipConfirmation   = flagSet.Bool("force", false, "skips user confirmation step allowing programmatic use")
+		apiFlags           = api.NewFlags(flagSet)
+	)
+
+	handler := func(args []string) error {
+		if err := flagSet.Parse(args); err != nil {
+			return err
+		}
+		if *daysToDelete < 60 {
+			fmt.Println("-days flag must be set to 60 or greater")
+			return nil
+		}
+
+		ctx := context.Background()
+		client := cfg.apiClient(apiFlags, flagSet.Output())
+
+		query := `
+query Users($first: Int, $query: String) {
+	users(first: $first, query: $query) {
+		nodes {
+			...UserFields
+		}
+	}
+}
+` + userFragment
+
+		// get users to delete
+		var result struct {
+			Users struct {
+				Nodes []User
+			}
+		}
+		if ok, err := client.NewRequest(query, nil).Do(ctx, &result); err != nil || !ok {
```

Comment:
@mrnugget this is great feedback. This request does always return all users, and that makes sense because we need to check every users activity for this. That being said I'd be way down to learn about altering the graphQL schema -- and just generally exploring the n+1 problem for the backend. This does need to be stable on very large instances as its being developed in direct relation to requests from such orgs 👍 _My naive understanding is that the client side computation should be too big a deal though, since its just basic dates and arithmetic?_

Will throw some time on to discuss with you further 👍

Reply:
> My naive understanding is that the client side computation shouldn't be too big a deal though, since its just basic dates and arithmetic?

True. The computation here is not the problem. It would be totally fine to compute this locally if you already had the data locally.

In this case, though, you first need to fetch the data. And if possible it's good to reduce amount of data being sent over the wire. Especially here, since you're loading _all users_ just to filter them (imagine a scenario in which you'd want to compute "average time since last activity" - for that you'd need to fetch all users locally).

On top of that: that particular query is slow, because I assume it's doing an N+1 query. The 1 query: it loads all users. The N queries: load usageStatistics for each users. So on an instance with 5000 users you end up running 5001 SQL queries, just to fetch the list of users and their `usageStatistics`. (I haven't looked at the code yet, so might be completely off).

[Try running this query on sourcegraph.com and see how long it takes.](https://sourcegraph.com/api/console#%7B%22query%22%3A%22query%20Users%20%7B%5Cn%20%20users%20%7B%5Cn%20%20%20%20nodes%20%7B%5Cn%20%20%20%20%20%20id%5Cn%20%20%20%20%20%20username%5Cn%20%20%20%20%20%20displayName%5Cn%20%20%20%20%20%20usageStatistics%20%7B%5Cn%20%20%20%20%20%20%20%20lastActiveTime%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%20%20%7D%5Cn%20%20%20%20pageInfo%20%7B%5Cn%20%20%20%20%20%20hasNextPage%5Cn%20%20%20%20%20%20endCursor%5Cn%20%20%20%20%7D%5Cn%20%20%20%20%5Cn%20%20%7D%5Cn%7D%5Cn%22%2C%22operationName%22%3A%22Users%22%7D)

We can fix the N+1. That might make the request faster. But I'd still think it'd nicer to do the filtering on the server and send less data. The database would have to do less work, the internet would have to transfer less data, the client would need to do less.

Let's use an extreme example to kinda highlight the pattern: say you want to build something like `src download-archives -repo-has-file=horse.txt` that downloads ZIP archives of all repositories that contain `horse.txt` files. You could download _all_ archives of all repositories and then check them locally and throw away the rest. Even ignoring that it'd take up disk space locally, that feels a bit wasteful, right? So instead we can move the compution to the data (instead of moving data to the computation) and ask our server to find repositories that contain this `horse.txt` file (since the server already has the files on disk) and only return those.